### PR TITLE
[Accessibility] Update show password icon to use a button element

### LINKED ISSUE #51825: Accessibility: CSS :before and :after pseudo-elements insert non-decorative content [My Account]
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

### Issue Description:

The CSS :before and :after pseudo-elements are used to add non-decorative content before and after an element's content in the document tree.

CSS :before and :after pseudo-elements specify the location of content before and after an element's document tree content. The content property, in conjunction with these pseudo-elements, specifies what is inserted.

Users who need to customize style information to view content according to their needs may not be able to access information inserted via CSS.

It is an accessibility failure to use these properties to insert non-decorative content.

T002_ShowHideIconDefinedThroughCSS

### Component or Page: 

Front-end Login Form on My Account Page

### Name affected element:

Show/Password icon

### Relevant code snippets:

```
<span class=""show-password-input""></span>

CSS:
.woocommerce form .show-password-input::after, .woocommerce-page form .show-password-input::after {
    font-family: WooCommerce;
    speak: never;
    font-weight: 400;
    font-variant: normal;
    text-transform: none;
    line-height: 1;
    -webkit-font-smoothing: antialiased;
    margin-left: .618em;
    content: ""\e010"";
    text-decoration: none;
}
```

### Recommended fix:

Ensure that CSS :before and :after pseudo-elements are not employed for inserting non-decorative content.

Replace this eye icon with an SVG or a font icon contained in the button rather then adding it via CSS :after.

### LINKED ISSUE #51826: Accessibility: Element is not keyboard-operable – <span> instead of buttons
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

### Issue Description:

The element(s) are not keyboard-focusable or clickable without a mouse because they are coded as <span> instead of HTML buttons.

T002_ShowHideIconIsNotReachableThroughKB

### Component or Page: 

Front-end Login Form on My Account Page

### Name affected element:

Show/Password icon

### Relevant code snippets:

`<span class="show-password-input"></span>`

### Recommended fix:

Instead of using the <span> element, it is recommended to use HTML <button> tags for these elements to make them inherently operable with keyboard.
 
 If you cannot replace a span with a button, it is possible to remediate it to function as a button with ARIA. The span would need to be given a button role and tabindex=""0"" to ensure that it can be focused with the keyboard. You would also need to write JavaScript to ensure that the span will function properly as a button for keyboard users. For more information on the aria button role, see: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role

### LINKED ISSUE #51827: Accessibility: Functional icon missing role and accessible name
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

### Issue Description:

The icon acts as a button but does not have the required role or an accessible name. Screen readers are skipping the icon. The icon is also not keyboard-focusable, meaning keyboard-only or screen reader users cannot access this functionality.

The button also does not have any way of communicating to screen reader users in it is currently pressed (or not).

T002_ShowHideIconDoNotRoleName

### Component or Page: 

Front-end Login Form on My Account Page

### Name affected element:

Show/Password icon

### Relevant code snippets:

`<span class="show-password-input"></span>`

### Recommended fix:

Interactive elements need to be coded with HTML <button> tags so that they can be focused and triggered with a keyboard alone. This element needs to be re-coded to be contained in a button tag. Ensure the button has an aria-label or screen reader only text that accurately describes its purpose.
 
 If you cannot change this to be contained in a <button> tag, it is possible to remediate a div or other container around the icon to function as a button with ARIA. The div would need to be given a button role and tabindex=""0"" to ensure that it can be focused with the keyboard. You would also need to write JavaScript to ensure that the div will function properly as a button for keyboard users and can be triggered with the space bar in addition to the return/enter key. For more information on turning a div into a button, see: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role
 
 This button also needs an aria-pressed attribute to communicate to screen reader users if their password is currently hidden or visible.

### LINKED ISSUE #51834: Accessibility: Element is not keyboard-operable – <span> instead of buttons [2]
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

### Issue Description:

The element(s) are not keyboard-focusable or clickable without a mouse because they are coded as <span> instead of HTML buttons.

T007_KB_MissingKB

### Component or Page: 

Account details My Account Page

### Name affected element:

Show/Password icon

### Relevant code snippets:

`<span class=""show-password-input""></span>`

### Recommended fix:

Instead of using the <span> element, it is recommended to use HTML <button> tags for these elements to make them inherently operable with keyboard.
 
 If you cannot replace a span with a button, it is possible to remediate it to function as a button with ARIA. The span would need to be given a button role and tabindex=""0"" to ensure that it can be focused with the keyboard. You would also need to write JavaScript to ensure that the span will function properly as a button for keyboard users. For more information on the aria button role, see: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role

### LINKED ISSUE #51835: Accessibility: Functional icon missing role and accessible name [My Account]
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

### Issue Description:

The icon acts as a button but does not have the required role or an accessible name. Screen readers are skipping the icon. The icon is also not keyboard-focusable, meaning keyboard-only or screen reader users cannot access this functionality.

The button also does not hav e any way of communicating to screen reader users in it is currently pressed (or not).

T007_KB_ShowHideIconMissingRoleName

### Component or Page: 

Account details My Account Page

### Name affected element:

Show/Password icon

### Relevant code snippets:

`<span class=""show-password-input""></span>`

### Recommended fix:

Interactive elements need to be coded with HTML <button> tags so that they can be focused and triggered with a keyboard alone. This element needs to be re-coded to be contained in a button tag. Ensure the button has an aria-label or screen reader only text that accurately describes its purpose.
 
 If you cannot change this to be contained in a <button> tag, it is possible to remediate a div or other container around the icon to function as a button with ARIA. The div would need to be given a button role and tabindex=""0"" to ensure that it can be focused with the keyboard. You would also need to write JavaScript to ensure that the div will function properly as a button for keyboard users and can be triggered with the space bar in addition to the return/enter key. For more information on turning a div into a button, see: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role
 
 This button also needs an aria-pressed attribute to communicate to screen reader users if their password is currently hidden or visible.

### PR DESCRIPTION
**Pull request by darrenjacoby**
### Submission Review Guidelines:

- * I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- * I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- * I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- * Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #51834  .
Closes #51826 .
Closes #51827 .
Closes #51825 .
Closes #51835 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Account:**

1. Browse to My Account → Account Details.
2. The Password Change form should display the Show/Hide Password icon as a `<button>` element.
3. The user should be able to tab to the Show/Hide Password button.
4. The button should show a focus state.
5. The button should have an `aria-label` that describes the action for screen readers.

**Login:**

1. Browse to My Account → Login.
2. The Password Change form should display the Show/Hide Password icon as a `<button>` element.
3. The user should be able to tab to the Show/Hide Password button.
4. The button should show a focus state.
5. The button should have an `aria-label` that describes the action for screen readers.

**Other dev notes:**

- I've updated the focus state once the Show/Hide Password button has been actioned to focus back to the input field. Let me know if this is considered favourable.
- I've added `aria-describedBy` for the Show/Hide Password button to link back to the relevant input field.
- I've updated themes Twenty Seventeen, Twenty Twenty and Twenty Twenty-One to avoid style regressions.
- Noting that Storefront could use `background-color: transparent` on the `.show-password-input` class. Do we have recommendation of where we would want to add that?

### Changelog entry

- * [ ] Automatically create a changelog entry from the details below.
- * [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Type

</details>

### CODE REVIEW

**Comment by darrenjacoby** [comment]
@dkotter ready for your review, left a few dev notes above, let me know your thoughts.

File: plugins/woocommerce/client/legacy/js/frontend/woocommerce.js
Comment (dkotter) [inline plugins/woocommerce/client/legacy/js/frontend/woocommerce.js]: I believe we'll want these `aria-labels` to be translated, whereas right now they are hardcoded to the English version

File: plugins/woocommerce/client/legacy/js/frontend/woocommerce.js
Comment (dkotter) [inline plugins/woocommerce/client/legacy/js/frontend/woocommerce.js]: Same comment on this label, it will need to be passed through translation

**Review by dkotter — CHANGES_REQUESTED** [review]

**Comment by darrenjacoby** [comment]
@dkotter good point, I've updated the PR to add `woocommerce.i18n_password_show` and `woocommerce.i18n_password_hide`. I'm not sure if this is a desirable structure (under namespace woocommerce), so let me know your thoughts.

**Comment by dkotter** [comment]
> @dkotter good point, I've updated the PR to add `woocommerce.i18n_password_show` and `woocommerce.i18n_password_hide`. I'm not sure if this is a desirable structure (under namespace woocommerce), so let me know your thoughts.

That works but also means that data will be added on all page views I believe. Ideally we only add this extra information on pages that need it (which seems like the My Account pages only). I'm not sure if there's a script that only loads on those pages that we could attach to? I see in the `get_script_data` method there's a `wc-password-strength-meter` script, could adding these values there work?

**Comment by darrenjacoby** [comment]
> > @dkotter good point, I've updated the PR to add `woocommerce.i18n_password_show` and `woocommerce.i18n_password_hide`. I'm not sure if this is a desirable structure (under namespace woocommerce), so let me know your thoughts.
> 
> That works but also means that data will be added on all page views I believe. Ideally we only add this extra information on pages that need it (which seems like the My Account pages only). I'm not sure if there's a script that only loads on those pages that we could attach to? I see in the `get_script_data` method there's a `wc-password-strength-meter` script, could adding these values there work?

Yea, that works too, and as you mentioned would fire only on the required pages. I've updated that code. The semantics aren't "exact", but we can rename `wc_password_strength_meter_params` if it's an issue.

**Review by dkotter — APPROVED** [review]
@ankitguptaindia @qasumitbagthariya This is ready for QA

**Comment by ankitguptaindia** [comment]
QA Status: Need fixes. ⚠️ 
----


@darrenjacoby Thank you for the fixes and PR! The changes are working as expected on the **My Account > Account Details page**, and the 'eye' icon now appears as a button.

However, I noticed that the eye icon no longer appears on the **Login page.** 

image


I tested this on both the specific branch with fixes and the trunk branch with different themes. Please check the video below for the steps I followed.


**Login page with fix specific branch:** 

https://github.com/user-attachments/assets/52de9313-4d80-4e9a-ac1c-8f63242f2d2f


**Login page with `trunk` branch:**  

https://github.com/user-attachments/assets/c8a8b160-cc65-43f8-a8cb-8087f8b936a1

**Comment by darrenjacoby** [comment]
Hi @ankitguptaindia , thanks for catching that. @dkotter it's a regression from moving the params, I've moved them back to `woocommerce` which feels correct as the referenced code lives within `woocommerce.js` as well. 

@ankitguptaindia , do you mind re-testing?

**Comment by ankitguptaindia** [comment]
**QA/Test Report-** ✅ 
------------

**Testing Environment -** 

<Details>

* WordPress: 6.7.1
* Theme:  
  - Storefront Version: 4.6.0  
  - Twenty Twenty-Four Version: 1.3  
  - Twenty Twenty-Five Version: 1.0
* WooCommerce - Development version with this fix-specific branch.  
* PHP: 8.3
* Web Server: Nginx
* Browser: Chrome - Version 131
* OS: macOS Sonoma 14.5


</Details>


**Test Results -** 

The changes work as intended and tested across Safari, Firefox, and Chrome. Accessibility was verified using macOS VoiceOver Version 10. Testing was conducted with the Twenty Twenty-Five, Storefront theme and Twenty Twenty-Four theme.
 

**Functional Demo /  Screencast -**

**With macOS Voice Over:**

https://github.com/user-attachments/assets/3d01a602-d3b6-4355-9050-a3470e6b1ae5

**Comment by ankitguptaindia** [comment]
Thanks for the great work on this PR @darrenjacoby would you mind updating this PR status to open for Woo CR? Thank you! 

cc: @vikrampm1

**Review by ankitguptaindia — APPROVED** [review]

**Comment by vikrampm1** [comment]
@mikejolley this PR is ready for your review!

**Review by mikejolley — APPROVED** [review]
This looks good to me. Tested across a few default themes. 

For storefront I think the fix to the color would need to be in that theme because it dequeues core stylesheets, unless we want to inline some basic styles.

**Comment by darrenjacoby** [comment]
@mikejolley yea, we should handle any storefront style changes as a separate PR on the storefront repo. It does however render quite well without any changes. I would just add `background-color: transparent` to clean it up a bit, but no functional regressions.

image

**Comment by vikrampm1** [comment]
@mikejolley if there are no further questions then please proceed and merge the PR, thanks!

**Comment by mikejolley** [comment]
There were some linting errors after running CI (https://github.com/woocommerce/woocommerce/pull/53489). Lints don't run on forks. I can take care of them assuming the rest of the CI passes.

**Comment by mikejolley** [comment]
@darrenjacoby @vikrampm1 Could you see this comment? https://github.com/woocommerce/woocommerce/pull/53489#pullrequestreview-2483802092 If this needs changing, please cherry pick the two linting fixes I made as well.

**Comment by vikrampm1** [comment]
@darrenjacoby this requires further updates before it can be merged, so please make the necessary changes and then notify Mike to merge the PR.

**Comment by darrenjacoby** [comment]
@mikejolley that is a legitimate flag, and something I noticed as well when testing. While this improves accessibility, it may impact user experience negatively. The user behaviour, based on the elements, is expected. It may require more discussions with the accessibility team and their recommendations in such cases. cc @vikrampm1

**Comment by vikrampm1** [comment]
@amberhinds do you have any comments on https://github.com/woocommerce/woocommerce/pull/53489#pullrequestreview-2483802092 as there is a concern of this PR, while resolves https://github.com/woocommerce/woocommerce/issues/51834, https://github.com/woocommerce/woocommerce/issues/51826, https://github.com/woocommerce/woocommerce/issues/51827, https://github.com/woocommerce/woocommerce/issues/51825 and https://github.com/woocommerce/woocommerce/issues/51835, may impact user experience.

**Comment by amberhinds** [comment]
RE @nielslange's comment
> Before this PR, a user could navigate to My Account → Login, tab into the username field, press Enter, provide their password, and press Enter again to log in. With this PR, pressing Enter after providing the password toggles the password visibility instead. As a result, pressing Enter no longer submits the form.

I agree that the loss of Return key functionality in the password field to submit the form may frustrate users who are used to it working this way. I'm not certain that it's a problem to remove that functionality, though - it's more of an annoyance. I asked our accessibility specialist for her opinion as well.

Is it possible to add it back and still address the password visibility button?

**Comment by amberhinds** [comment]
We discussed this, and removing the ability to submit the form by pressing return in the password field does not create an accessibility issue, so if it's not possible to add that functionality back, I don't think it's a problem. It's better to have a keyboard-accessibility button for password visibility.

**Comment by darrenjacoby** [comment]
@amberhinds @mikejolley @nielslange let me know if the implementation above could work.

It disables the "Enter" key for the password field (I don't think we should show the password if the user presses "Enter", it feels like it could pose an security risk). For the login form, pressing Enter on the password field triggers a pseudo event of clicking "Submit".

Please let me know your thoughts if this feels viable.

**Review by nielslange — APPROVED** [review]
I’ve tested all the changes successfully, @darrenjacoby. Thanks again for your work on this! I'll merge the PR once all CI checks have passed.

Regarding https://github.com/woocommerce/woocommerce/pull/53098#issuecomment-2519754633, could you create an issue in https://github.com/woocommerce/storefront/issues?

File: plugins/woocommerce/includes/class-wc-frontend-scripts.php
Comment (nielslange) [inline plugins/woocommerce/includes/class-wc-frontend-scripts.php]:
```suggestion
				$params = array(
					'ajax_url'           => WC()->ajax_url(),
					'wc_ajax_url'        => WC_AJAX::get_endpoint( '%%endpoint%%' ),
					'i18n_password_show' => esc_attr__( 'Show password', 'woocommerce' ),
					'i18n_password_hide' => esc_attr__( 'Hide password', 'woocommerce' ),
				);
```

Comment (nielslange) [reply plugins/woocommerce/includes/class-wc-frontend-scripts.php]: While running the CI checks, I noticed the following lint error.

**Comment by amberhinds** [comment]
Sorry, I'm late to the game. This seems like a good solution to me. I agree that unexpectedly showing someone's password could be a security issue.

**Pull request merged**
